### ISSUE TITLE
Trigger force sync

### ISSUE THREAD

**Issue by asiripanich** [e-mission/e-mission-docs#529]
@shankari is it possible to trigger force sync every time on emTripLog the user completes a trip-end survey? (embedded Enketo survey)

We found that survey responses never got sent to the server before the next trip was finished.

**Comment by asiripanich** [comment]
@atton16

**Comment by asiripanich** [comment]
@shankari any advice on this?

**Comment by shankari** [comment]
@asiripanich 

> We found that survey responses never got sent to the server before the next trip was finished.

This has been true for a very long time - e.g.
https://github.com/e-mission/e-mission-docs/issues/281#issuecomment-446076440

Force sync will not help because we currently only push data when we detect a completed trip. Why does it matter if the survey is sent only when the next trip was finished?

**Comment by asiripanich** [comment]
For our study, we created a procedure that our participants have to follow. The last step of the procedure requires answering to a post-experience (PE) survey. Once the survey is answered, they are allowed to uninstall the app. Hence, any locally stored data won't be sent if the users delete the app immediately after they have answered the PE survey.

**Comment by shankari** [comment]
didn't you already ask this question before?
https://github.com/e-mission/e-mission-docs/issues/509

The solution is simple - you have to simulate a trip start and end before you force sync. I already have a function for this in the control code - you should invoke that.

Your original question was about trip end surveys, where simulating a trip start and end would be more tricky. But this is the perfect solution for a study end survey.

**Comment by asiripanich** [comment]
;) neat. Thanks Shankari. 

> Your original question was about trip end surveys, where simulating a trip start and end would be more tricky. 

If a solution exists we would like to do that as well.

**Comment by asiripanich** [comment]
@atton16 mentioned that you once told him emTripLog will only send data to the server if it has enough data to be sent? is this true? In that case, will a fake trip that gets created with the force sync function will be enough?

**Comment by asiripanich** [comment]
> didn't you already ask this question before?
> 
> https://github.com/e-mission/e-mission-docs/issues/509
> 
> 
> 
> The solution is simple - you have to simulate a trip start and end before you force sync. I already have a function for this in the control code - you should invoke that.
> 
> 
> 
> Your original question was about trip end surveys, where simulating a trip start and end would be more tricky. But this is the perfect solution for a study end survey.

Btw, shouldn’t there be a more explicit way to do this?

**Comment by shankari** [comment]
> In that case, will a fake trip that gets created with the force sync function be enough?

Yes. @atton16 misinterpreted my comment. e-mission does not look for "enough data to be sent". It looks for a completed trip. Typically a completed trip is associated with data

As I said, there is a button on the control screen that does this. Use the same function that it does and you will be fine.

> Btw, shouldn’t there be a more explicit way to do this?

Why? In general, for long-term use, there are no guarantees that data will be pushed to the server within any particular time period. The data is cached locally, and will stay cached until it is pushed. The user could turn cellular data off for the app and then the trips will only be pushed when they next have access to WiFi. Once they do have access, we push the data in chunks of 10000 entries as a time, so not all the data might make it in one round.

If you can articulate a specific feature request and the reason for it, that would be good.

**Comment by shankari** [comment]
Fixed by https://github.com/e-mission/e-mission-phone/pull/660, https://github.com/e-mission/e-mission-phone/pull/661, https://github.com/e-mission/e-mission-phone/pull/662

**Comment by asiripanich** [comment]
> > In that case, will a fake trip that gets created with the force sync function be enough?
> 
> Yes. @atton16 misinterpreted my comment. e-mission does not look for "enough data to be sent". It looks for a completed trip. Typically a completed trip is associated with data
> 
> As I said, there is a button on the control screen that forces a trip start/end and then a forceSync. Use the same function that it does and you will be fine.
> 
> > Btw, shouldn’t there be a more explicit way to do this?
> 
> Why? In general, for long-term use, there are no guarantees that data will be pushed to the server within any particular time period. The data is cached locally, and will stay cached until it is pushed. The user could turn cellular data off for the app and then the trips will only be pushed when they next have access to WiFi. Once they do have access, we push the data in chunks of 10000 entries as a time, so not all the data might make it in one round.
> 
> If you can articulate a specific feature request and the reason for it, that would be good.

Does it mean that a chuck has to be filled with at least 10000 entries to get pushed to the server?

**Comment by asiripanich** [comment]
@shankari 

In our study, currently active, we send out notifications to all our users to remind them of their unanswered trip-end questions. Since we have to process this from the server side we need to have the most update-to-date data to notify our users correctly. So if we can have a proper way, instead of simulating a zero movement trip, to push locally cached data to the server to be processed, we can send out the notifications correctly. 

![UNADJUSTEDNONRAW_thumb_1431](https://user-images.githubusercontent.com/17020181/82327172-5a6e5580-9a21-11ea-95ea-2fe2f1e6e20c.jpg)

**Comment by shankari** [comment]
> Does it mean that a chuck has to be filled with at least 10000 entries to get pushed to the server? even with forceSynce?

No, it means that if there are more than 10,000 entries, then they will be pushed 10,000 at a time. So if you have 50,000 cached entries, then it will take 5 syncs to get them to the server. If you have less than 10,000 entries (the common case), they will be pushed on every trip end.

**Comment by shankari** [comment]
> In our study, currently active, we send out notifications to all our users to remind them of their unanswered trip-end questions. 

Just to clarify, any fixes to this will not affect your current study since they will require changes to native code. But such feedback will make the platform better for future uses!

**Comment by shankari** [comment]
> In our study, currently active, we send out notifications to all our users to remind them of their unanswered trip-end questions. 

Just to clarify again, since all the data is pushed at the end of a trip, this is off by at most one for the last day. So in the screenshot above, if the feature that you request was implemented, the 2h ago notification would say
"There are 3 trips from 3,7,19 of May that haven't been answered".

Does that match your understanding?
